src/randomness/hash_chain.rs

```diff
@@ -0,0 +1,243 @@
+                &mut random_bytes_out[offset..offset + KECCAK256_DIGEST_NUM_BYTES],
+            );
+        }
+
+        let num_tail_bytes = num_bytes % KECCAK256_DIGEST_NUM_BYTES;
+        if num_tail_bytes <= self.num_spare_bytes {
+            random_bytes_out[num_full_blocks * KECCAK256_DIGEST_NUM_BYTES..num_bytes]
+                .copy_from_slice(&self.spare_bytes[..num_tail_bytes]);
+            self.num_spare_bytes -= num_tail_bytes;
+
+            self.spare_bytes.copy_within(num_tail_bytes.., 0);
+        } else {
+            self.fill_random_bytes(
+                &mut random_bytes_out[num_full_blocks * KECCAK256_DIGEST_NUM_BYTES..num_bytes],
+            );
+        }
+    }
+
+    fn fill_random_bytes(&mut self, out: &mut [u8]) {
+        let num_bytes: usize = out.len();
+        assert!(
+            num_bytes <= KECCAK256_DIGEST_NUM_BYTES,
+            "Asked to get more bytes than one digest size"
+        );
+
+        let prandom_bytes = self.next_hash();
+        out.copy_from_slice(&prandom_bytes[..num_bytes]);
+
+        assert!(
+            self.num_spare_bytes < KECCAK256_DIGEST_NUM_BYTES + num_bytes,
+            "Not enough room in spare bytes buffer. Have {} bytes and want to add {} bytes",
+            self.num_spare_bytes,
+            KECCAK256_DIGEST_NUM_BYTES - num_bytes
+        );
+
+        self.spare_bytes
+            [self.num_spare_bytes..self.num_spare_bytes + (KECCAK256_DIGEST_NUM_BYTES - num_bytes)]
+            .copy_from_slice(&prandom_bytes[num_bytes..]);
+        self.num_spare_bytes += KECCAK256_DIGEST_NUM_BYTES - num_bytes;
+        self.counter += 1;
+    }
+
+    fn next_hash(&self) -> [u8; KECCAK256_DIGEST_NUM_BYTES] {
+        // TODO: below code is not efficient, but it works for now
+        let mut bytes_with_counter = [0u8; KECCAK256_DIGEST_NUM_BYTES * 2];
+        bytes_with_counter[..KECCAK256_DIGEST_NUM_BYTES].copy_from_slice(&self.digest);
+
+        // Copy the counter's serialized 64bit onto the MSB end of the buffer (stone-prover PR #875 decision).
+        // Serialized counter is in big-endian format.
+        bytes_with_counter[KECCAK256_DIGEST_NUM_BYTES * 2 - 8..]
+            .copy_from_slice(&self.counter.to_be_bytes());
+
+        let mut hasher = Keccak256::new();
+        hasher.update(bytes_with_counter);
+        let result = hasher.finalize();
+        let mut hash_bytes = [0u8; KECCAK256_DIGEST_NUM_BYTES];
+        hash_bytes.copy_from_slice(&result);
+        hash_bytes
+    }
+
+    pub fn update_hash_chain(&mut self, raw_bytes: &[u8]) {
+        let seed_increment: u64 = 0;
+        self.mix_seed_with_bytes(raw_bytes, seed_increment);
+    }
+
+    pub fn mix_seed_with_bytes(&mut self, raw_bytes: &[u8], seed_increment: u64) {
+        let mut mixed_bytes = vec![0u8; KECCAK256_DIGEST_NUM_BYTES + raw_bytes.len()];
+
+        // Deserialize the current digest into a u64 array
+        let big_int = U256::from_be_bytes(self.digest).add(U256::from(seed_increment));
```

Comment:
If all we need is an addition maybe we should just implement that ourselves:

Adding the low 64-bits and carrying forward the overflow should only be a couple lines.

Reply:
I used ethnum U256 because there's no from_be_bytes method in ark_ff::BigInteger. There's only from_bits_be 😓. I thought all the process of changing bytes into bits and then... is kind of redundant.
Implementing ADC ourselves seems to be the best choice out of all options. 👍